### DISCUSSION TITLE
edgedb-python: access link property, with typing

### DISCUSSION THREAD

**Discussion by fantix** [geldata/gel#4544] [Pre-RFC]
The problem is: current (v0.24) way to access link property is tricky for historical reasons. We are trying to simplify the access to link properties in edgedb-python. At the same time, it would be best if link properties could get typing support in `codegen` generated code.

Schema:

```
type Target { ... }

type Object {
    required property attr -> int32;
    required link "link" -> Target {
        required property lp -> int32;
    }
}
```

Generated code is like:

```python
@dataclasses.dataclass
class Target:
    ...

@dataclasses.dataclass
class Object:
    attr: int
    link: Target
```

1. In current v0.24, `obj.link` returns a `Target` instance, while `obj['link']` returns a `edgedb.Link` instance, which has the link property and accessible through `obj['link'].lp`.
2. edgedb/edgedb-python#379 proposes to deprecate the `edgedb.Link` instruct, and allow accessing link properties through `obj.link['@lp']`.
3. @1st1  suggested for `v1.0`, we could also allow accessing regular properties like `obj['attr']`, but this gets confusing when it comes to `obj['link']`.
4. v2.0 is just to say when we might drop the deprecated APIs.
5. "typing" column is for `codegen` generated code, I am assuming we cannot type square bracket values based on the keys.
6. @1st1 proposed one possible option for typing the link properties in scenario (18) `obj.link.__linkprop__.lp`, but it is kinda long. So we're willing to hear some python/mypy black magic to address the typing problem nicely.

|    | exp                        | 0.24              | edgedb/edgedb-python#379                          | 1.0                                               | 2.0                                       | typing |
|----|----------------------------|-------------------|---------------------------------------------------|---------------------------------------------------|-------------------------------------------|--------|
| 1  | `obj.attr`                 | `int`             | `int`                                             | `int`                                             | `int`                                     | yes    |
| 2  | `getattr(obj, 'attr')`     | `int`             | `int`                                             | `int`                                             | `int`                                     | yes    |
| 3  | `obj['attr']`              | `TypeError`       | `TypeError`                                       | maybe `int`? :pencil2:                            |                                           | no     |
| 4  | `obj['@attr']`             | `KeyError` (link) | `KeyError` (link prop)                            | `KeyError` (link prop) ?                          |                                           | -      |
| 5  | `getattr(obj, '@attr')`    | `AttributeError`  | `AttributeError`                                  | `AttributeError`                                  | `AttributeError`                          | -      |
| 6  | `obj.link`                 | `Target`          | `Target`                                          | `Target`                                          | `Target`                                  | yes    |
| 7  | `getattr(obj, 'link')`     | `Target`          | `Target`                                          | `Target`                                          | `Target`                                  | yes    |
| 8  | `obj['link']`              | `edgedb.Link`     | `edgedb.Link` with `DeprecationWarning` :pencil2: | `edgedb.Link` with `DeprecationWarning` :pencil2: | `KeyError` or `Target`? :pencil2:         | no     |
| 9  | `obj['@link']`             | `KeyError` (link) | `KeyError` (link prop)                            | `KeyError` (prop)?                                |                                           | -      |
| 10 | `getattr(obj, '@link')`    | `AttributeError`  | `AttributeError`                                  | `AttributeError`                                  | `AttributeError`                          | -      |
| 11 | `obj.link.lp`              | `AttributeError`  | `AttributeError`                                  | `AttributeError`                                  | `AttributeError`                          | -      |
| 12 | `getattr(obj.link, 'lp')`  | `AttributeError`  | `AttributeError`                                  | `AttributeError`                                  | `AttributeError`                          | -      |
| 13 | `obj.link['lp']`           | `KeyError` (link) | `TypeError` (@ expected)                          | `KeyError` (prop)?                                |                                           | no     |
| 14 | `obj.link['@lp']`          | `KeyError` (link) | `int` :pencil2:                                   | `int` :pencil2:                                   | `int` :pencil2:                           | no     |
| 15 | `getattr(obj.link, '@lp')` | `AttributeError`  | `AttributeError`                                  | `int`? :pencil2:                                  |                                           | no     |
| 16 | `obj['link'].lp`           | `int`             | `int` with `DeprecationWarning` :pencil2:         | `int` with `DeprecationWarning` :pencil2:         | `KeyError` or `AttributeError`? :pencil2: | no     |
| 17 | `obj['link']['lp']`        | `KeyError` (link) | `KeyError` (link prop)                            | `KeyError` (@ expected)?                          | `KeyError`                                | -      |
| 18 | `obj.link.__linkprop__.lp` |                   |                                                   | `int` :pencil2:                                   |                                           | yes    |

**Comment by msullivan** [comment]
> "typing" column is for codegen generated code, I am assuming we cannot type square bracket values based on the keys.

It is actually totally doable to type square brackets based on keys, using overloads and literal types:
```
from typing import *

class X:
    @overload
    def __getitem__(self, key: Literal['foo']) -> str:
        ...

    @overload
    def __getitem__(self, key: Literal['bar']) -> int:
        ...

    def __getitem__(self, key: str) -> Any:
        raise NotImplementedError


x = X()
reveal_type(x['foo'])
reveal_type(x['bar'])
```
gives
```
test.py:17:13: note: Revealed type is "builtins.str"
test.py:18:13: note: Revealed type is "builtins.int"
```

**Comment by msullivan** [comment]
Another option to consider is adding a *method* to access link properties. (It could also use overloads to have types, in generated code.)

**Comment by fantix** [comment]
Some decisions:

1. We'll use (14) to access link properties.
2. We don't want to access regular properties or links with square brackets yet.

|    | exp                        | 0.24              | edgedb/edgedb-python#379                          | 1.0                                               | 2.0                     | typing |
|----|----------------------------|-------------------|---------------------------------------------------|---------------------------------------------------|-------------------------|--------|
| 1  | `obj.attr`                 | `int`             | `int`                                             | `int`                                             | `int`                   | yes    |
| 2  | `getattr(obj, 'attr')`     | `int`             | `int`                                             | `int`                                             | `int`                   | yes    |
| 3  | `obj['attr']`              | `TypeError`       | `TypeError`                                       | `TypeError`                                       | `TypeError`             | -      |
| 4  | `obj['@attr']`             | `KeyError` (link) | `KeyError` (link prop)                            | `KeyError` (link prop)                            | `KeyError` (link prop)  | -      |
| 5  | `getattr(obj, '@attr')`    | `AttributeError`  | `AttributeError`                                  | `AttributeError`                                  | `AttributeError`        | -      |
| 6  | `obj.link`                 | `Target`          | `Target`                                          | `Target`                                          | `Target`                | yes    |
| 7  | `getattr(obj, 'link')`     | `Target`          | `Target`                                          | `Target`                                          | `Target`                | yes    |
| 8  | `obj['link']`              | `edgedb.Link`     | `edgedb.Link` with `DeprecationWarning` :pencil2: | `edgedb.Link` with `DeprecationWarning` :pencil2: | `KeyError` :pencil2:    | yes    |
| 9  | `obj['@link']`             | `KeyError` (link) | `KeyError` (link prop)                            | `KeyError` (link prop)                            | `KeyError` (link prop)  | -      |
| 10 | `getattr(obj, '@link')`    | `AttributeError`  | `AttributeError`                                  | `AttributeError`                                  | `AttributeError`        | -      |
| 11 | `obj.link.lp`              | `AttributeError`  | `AttributeError`                                  | `AttributeError`                                  | `AttributeError`        | -      |
| 12 | `getattr(obj.link, 'lp')`  | `AttributeError`  | `AttributeError`                                  | `AttributeError`                                  | `AttributeError`        | -      |
| 13 | `obj.link['lp']`           | `KeyError` (link) | `TypeError` (@ expected)                          | `KeyError` (@ expected)                           | `KeyError` (@ expected) | -      |
| 14 | `obj.link['@lp']`          | `KeyError` (link) | `int` :pencil2:                                   | `int` :pencil2:                                   | `int` :pencil2:         | yes    |
| 15 | `getattr(obj.link, '@lp')` | `AttributeError`  | `AttributeError`                                  | `int`? :pencil2:                                  |                         | no     |
| 16 | `obj['link'].lp`           | `int`             | `int` with `DeprecationWarning` :pencil2:         | `int` with `DeprecationWarning` :pencil2:         | `KeyError` :pencil2:    | no     |
| 17 | `obj['link']['lp']`        | `KeyError` (link) | `KeyError` (link prop)                            | `KeyError` (link prop)                            | `KeyError`              | -      |
